Feature: complete lifecycle signals and documentation (#1147)
pdm-project/pdm: docs/docs/usage/hooks.md

of a Python project lifecycle and PDM provides commands to perform the expected tasks for those phases.

It also provides hooks attached to these steps allowing for:

- plugins to listen to the [signals](../plugin/reference.md#signals) of the same name.
- developpers to define custom scripts with the same name.

The built-in commands are currently splitted in 3 groups:

- the [initialization phase](#initialization)
- the [dependencies management](#dependencies-managment).
- the [publication phase](#publication).

You will most probably need to perform some recurrent tasks between th installation and publication phases (housekeeping, linting, testing, ...)
this is why PDM lets you define your own tasks/phases using [user scripts](#user-scripts).

To provides full flexibility, PDM allows to [skip some hooks and tasks](#skipping) on demand.

## Initialization

The intialization phase should occurs only once in a project lifetime by running on of the following commands:

- `pdm new` to start a project from scratch
- `pdm init` to initialize an existing project (prompt to fill the `pyproject.toml` file)

They trigger the following hooks:

> - [`post_init`](../plugin/reference.md#pdm.signals.post_init)

Reviewer: You can take advantage of the autorefs MkDocs plugin (which is used by mkdocstrings) and write ``[`post_init`][pdm.signals.post_init]``. The advantage is you don't have to specify the page, so the link will work even if the structure of the docs changes.

Author: Way better 👌🏼

- `add`: add a dependency
- `remove`: remove a dependency

All thoses steps are directly available with the following commands:

- [`pdm lock`](dependency.md#install-the-packages-pinned-in-lock-file): execute the `lock` task
- [`pdm sync`](dependency.md#install-the-packages-pinned-in-lock-file): execute the `sync` task
- [`pdm install`](dependency.md#install-the-packages-pinned-in-lock-file): execute the `install` task, preceded from `lock` and/or `sync` if required
- [`pdm add`](dependency.md#add-dependencies): add a dependency requirement, re-lock and then sync
- [`pdm remove`](dependency.md#remove-existing-dependencies): remove a dependency requirement, re-lock and then sync
- [`pdm update`](dependency.md#update-existing-dependencies): re-lock dependencies from their latest versions and then sync

They trigger the following hooks:

- [`pre_install`](../plugin/reference.md#pdm.signals.pre_install)
- [`post_install`](../plugin/reference.md#pdm.signals.post_install)
- [`pre_lock`](../plugin/reference.md#pdm.signals.pre_lock)
- [`post_lock`](../plugin/reference.md#pdm.signals.post_lock)
- [`pre_sync`](../plugin/reference.md#pdm.signals.pre_sync)

> - [`post_sync`](../plugin/reference.md#pdm.signals.post_sync)

Reviewer: Same comment as above. Just a suggestion of course :slightly_smiling_face:

``` mermaid
flowchart LR
  subgraph pdm-install [pdm install]
    direction LR
    pre-install{{Emit pre_install}}
    post-install{{Emit post_install}}

    subgraph pdm-lock [pdm lock]
      direction TB
      pre-lock{{Emit pre_lock}}
      post-lock{{Emit post_lock}}
      pre-lock --> lock --> post-lock
    end
    
    subgraph pdm-sync [pdm sync]
      direction TB
      pre-sync{{Emit pre_sync}}
      post-sync{{Emit post_sync}}
      pre-sync --> sync --> post-sync
    end
    
    pre-install --> pdm-lock --> pdm-sync --> install --> post-install
  end
```

## Publication

As soon as you are ready to publish you package/app/libray, you will required the publish flow:

- `build`: build/compile assets requiring it
- `package`: package you library/application into a Python package
- `upload`: upload/publish the package to a remote PyPI index

All thoses steps are available with the following commands:

- [`pdm build`](project.md#build-distribution-artifacts)

> - [`pdm publish`](project/#build-distribution-artifacts)

Reviewer: This made me notice a discrepancy: `(page.md#...)` versus `(page/#...)`. I find the latter more readable.

Author: But this version was not working, it was missing the start of the path.
If finally chose the first one because of it: `.md` links are resolved by mkdocs and so survive URL changes and support proper sibling files references , while the second one forced you to know the expected final path.

# Lifecycle and hooks

As any Python deliverable, your project will go through the different phases
of a Python project lifecycle and PDM provides commands to perform the expected tasks for those phases.

It also provides hooks attached to these steps allowing for:

> - plugins to listen to the [signals][pdm.signals] of the same name.

Reviewer 2: Is this a link ref or something? I can't find it

Reviewer: More info: https://mkdocstrings.github.io/usage/#cross-references-to-any-markdown-heading (and previous/next paragraphs).

Reviewer 2: That is awesome.

- developpers to define custom scripts with the same name.

The built-in commands are currently splitted in 3 groups:

- the [initialization phase](#initialization)
- the [dependencies management](#dependencies-managment).
- the [publication phase](#publication).

You will most probably need to perform some recurrent tasks between the installation and publication phases (housekeeping, linting, testing, ...)
this is why PDM lets you define your own tasks/phases using [user scripts](#user-scripts).

To provides full flexibility, PDM allows to [skip some hooks and tasks](#skipping) on demand.

## Initialization

> The intialization phase should occurs only once in a project lifetime by running the [`pdm init`](cli_reference.md#exec-0--init) 

Reviewer 2: ```suggestion
The intialization phase should occur only once in a project lifetime by running the [`pdm init`](cli_reference.md#exec-0--init) 
```

command to initialize an existing project (prompt to fill the `pyproject.toml` file).

They trigger the following hooks:

- [`post_init`][pdm.signals.post_init]

``` mermaid
flowchart LR
  subgraph pdm-init [pdm init]
    direction LR
    post-init{{Emit post_init}}
    init --> post-init
  end 
```

## Dependencies management

> The dependencies management is required for the developper to be able to work and perform the following:

Reviewer 2: ```suggestion
The dependencies management is required for the developer to be able to work and perform the following:
```

- [`pdm add`](cli_reference.md#exec-0--add): add a dependency requirement, re-lock and then sync
- [`pdm remove`](cli_reference.md#exec-0--remove): remove a dependency requirement, re-lock and then sync
- [`pdm update`](cli_reference.md#exec-0--update): re-lock dependencies from their latest versions and then sync

They trigger the following hooks:

- [`pre_install`][pdm.signals.pre_install]
- [`post_install`][pdm.signals.post_install]
- [`pre_lock`][pdm.signals.pre_lock]
- [`post_lock`][pdm.signals.post_lock]

``` mermaid
flowchart LR
  subgraph pdm-install [pdm install]
    direction LR

    subgraph pdm-lock [pdm lock]
      direction TB
      pre-lock{{Emit pre_lock}}
      post-lock{{Emit post_lock}}
      pre-lock --> lock --> post-lock
    end
    
    subgraph pdm-sync [pdm sync]
      direction TB
      pre-install{{Emit pre_install}}
      post-install{{Emit post_install}}
      pre-install --> sync --> post-install
    end
    
    pdm-lock --> pdm-sync
  end
```

## Publication

As soon as you are ready to publish you package/app/libray, you will require the publication tasks:

- `build`: build/compile assets requiring it

> - `package`: package you library/application into a Python package

Reviewer 2: `package` may be a bit confusing here. In Python packaging world, `build` means the process the build system reads the configuration and produces sdist or wheel. It should cover the `package` phase. Also in PDM there are only `build` and `upload`, no such a phase named `package`.

Author: I merged it into the `build` description. I had been tricked by the `pdm.cli.publish.package` module

      pre-script --> pre-test --> test --> post-test --> post-script
    end

    pre-run --> run-test --> post-run
  end
```

while `pdm run composite` will have the following:

``` mermaid
flowchart LR
  subgraph pdm-run-composite [pdm run composite]
    direction LR
    pre-run{{Emit pre_run}}
    post-run{{Emit post_run}}

    subgraph run-composite [composite task]
      direction TB
      pre-script-composite{{Emit pre_script}}
      post-script-composite{{Emit post_script}}
      pre-composite[Execute pre_composite]
      post-composite[Execute post_composite]

      subgraph run-test [test task]
        direction TB
        pre-script-test{{Emit pre_script}}
        post-script-test{{Emit post_script}}
        pre-test[Execute pre_test]
        post-test[Execute post_test]

        pre-script-test --> pre-test --> test --> post-test --> post-script-test
      end

      pre-script-composite --> pre-composite --> run-test --> post-composite --> post-script-composite
    end

     pre-run --> run-composite --> post-run
  end
```

## Skipping

It is possible to control which task and hook runs for any built-in command as well as custom user scripts
using the `--skip` option.

> If accept a comma-separated list of hooks/task names to skip

Reviewer 2: ```suggestion
It accepts a comma-separated list of hooks/task names to skip
```

as well as the predefined `:all`, `:pre` and `:post` shortcuts
respectively skipping all hooks, all `pre_*` hooks and all `post_*` hooks.

> You can also provide you skip list in `PDM_SKIP_HOOKS` environment variable

Reviewer 2: ```suggestion
You can also provide the skip list in `PDM_SKIP_HOOKS` environment variable
```
